src/goto.jl

```diff
@@ -109,41 +113,54 @@ end
 ## module goto
 
 function GotoItem(mod::Module)
+  name = string(mod)
   file, line = mod == Main ? MAIN_MODULE_LOCATION[] : moduledefinition(mod)
-  GotoItem(string(mod), file, line - 1)
+  GotoItem(name, file, line - 1)
 end
 
 ## toplevel goto
 
-const PathItemsMaps = Dict{String, Vector{ToplevelItem}}
+const PathItemsMap = Dict{String, Vector{GotoItem}}
+
+# NOTE:
+# this data structure should only keep "simple" data so that it doesn't eat
+# too much memory when the cache has got fat (like via `regeneratesymbols`)
+# TODO?:
+# maybe we will end up creating a more generic struct that represents a "symbol",
+# and wanting `SYMBOLSCACHE` to store them instead of `GotoItem`
+# (should be kept "simple" atst though, and ideally it handles both toplevel and local symbols)
+# so that we can use `SYMBOLSCACHE` as a general cache interface for various static features
+# like goto, find references, and so on.
 
 """
     Atom.SYMBOLSCACHE
 
-"module" (`String`) ⟶ "path" (`String`) ⟶ "symbols" (`Vector{ToplevelItem}`) map.
+"module" (`String`) ⟶ "path" (`String`) ⟶ "symbols" (`Vector{GotoItem}`) map.
 
 !!! note
     "module" should be canonical, i.e.: should be identical to names that are
       constructed from `string(mod::Module)`.
 """
-const SYMBOLSCACHE = Dict{String, PathItemsMaps}()
+const SYMBOLSCACHE = Dict{String, PathItemsMap}()
 
 function toplevelgotoitems(word, mod, path, text)
   key = string(mod)
-  pathitemsmaps = if haskey(SYMBOLSCACHE, key)
+  pathitemsmap = if haskey(SYMBOLSCACHE, key)
     SYMBOLSCACHE[key]
   else
     SYMBOLSCACHE[key] = collecttoplevelitems(mod, path, text) # caching
   end
 
   ismacro(word) && (word = lstrip(word, '@'))
   ret = []
-  for (path, items) in pathitemsmaps
-    for item in filter(item -> filtertoplevelitem(word, item), items)
-      push!(ret, GotoItem(path, item))
-    end
+  for (_, items) in pathitemsmap
+    @>> filter(items) do item
+      let item = item
+        word == item.name
+      end
+    end append!(ret)
```

Comment:
sorry for this `@>>` macro, but I couldn't come up with the appropriate name for this